torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -0,0 +1,300 @@
+(see NOTE [ Version Counter Sharing ] in the pytorch repo). However, we can
+also do so with the new ``_unsafe_preserve_version_counter`` context, avoiding
+the ``.data`` usage and being more explicit with our intention. Thus, this is
+the approach that we take.
+"""
+
+
+class AllGatherResult(NamedTuple):
+    all_gather_output: torch.Tensor
+    all_gather_event: Optional[torch.cuda.Event]
+    all_gather_work: Optional[dist.distributed_c10d.Work]
+
+
+@torch.no_grad()
+def foreach_all_gather(
+    fsdp_params: List[FSDPParam],
+    group: dist.ProcessGroup,
+    async_op: bool,
+    all_gather_copy_in_stream: torch.cuda.Stream,
+    all_gather_stream: torch.cuda.Stream,
+    use_uint8: bool,
+    device: torch.device,
+) -> Optional[AllGatherResult]:
+    """
+    Args:
+        use_int8 (bool): If ``False``, then this requires that all parameters
+            have the same dtype, which is used for the all-gather. If ``True``,
+            then this permits different dtypes and converts everything to uint8
+            for the all-gather.
+
+    Returns:
+        Optional[AllGatherResult]: The all-gathered output to be copied out, a
+        CUDA event recording the end of the all-gather collective, and
+        optionally an all-gather work object if this all-gather is run as an
+        async op (for explicit prefetching). If there are no parameters that
+        need to be all-gathered, then returns ``None``.
+    """
+    fsdp_params = [
+        fsdp_param
+        for fsdp_param in fsdp_params
+        if fsdp_param.state != ShardedState.UNSHARDED
+    ]
+    if len(fsdp_params) == 0:
+        return None
+    # Already checked at construction time for a uniform unsharded parameter
+    # dtype if not using uint8
+    dtype = torch.uint8 if use_uint8 else fsdp_params[0].unsharded_param_data_dtype
+    world_size = group.size()
+    group_rank = group.rank()
+    if use_uint8:
+        total_padded_unsharded_numel = sum(
+            fsdp_param.padded_unsharded_bytes for fsdp_param in fsdp_params
+        )
+    else:
+        total_padded_unsharded_numel = sum(
+            fsdp_param.padded_unsharded_numel for fsdp_param in fsdp_params
+        )
+    total_sharded_numel = total_padded_unsharded_numel // world_size
+    with torch.cuda.stream(all_gather_copy_in_stream):
+        # - Copy in
+        all_gather_output = torch.empty(
+            (total_padded_unsharded_numel,), dtype=dtype, device=device
+        )
+        all_gather_input = all_gather_output.narrow(
+            0, total_sharded_numel * group_rank, total_sharded_numel
+        )
+        shards_to_cat = [fsdp_param.all_gather_input for fsdp_param in fsdp_params]
+        if use_uint8:
+            shards_to_cat = [t.view(torch.uint8) for t in shards_to_cat]
+        # TODO: If foreach copy supports different src/dst dtypes, then we
+        # should use that for the mixed precision case for fusion.
+        torch.cat(shards_to_cat, out=all_gather_input)
```

Comment:
If we have https://github.com/pytorch/pytorch/issues/115171, then we can do something like:
```
        if dtype != shards_to_cat[0].dtype:
            dsts = torch.split(
                all_gather_input, [t.numel() for t in shards_to_cat], dim=0
            )
            torch._foreach_copy_(dsts, shards_to_cat)
        else:
            torch.cat(shards_to_cat, out=all_gather_input)
```
This will also help the reduce-scatter foreach copy in the case like `param_dtype=bf16` and `reduce_dtype=fp32`.